src/main/java/io/github/townyadvanced/townyprovinces/jobs/map_display/DisplayProvincesOnBlueMapAction.java

```diff
@@ -0,0 +1,241 @@
+				}
+				for (BlueMapMap map : world.get().getMaps()) {
+					map.getMarkerSets().put("townyprovinces.markersets.borders", borderMarkerSet);
+				}
+			}
+
+			if (homeBlocksRefreshRequested) {
+				if (homeBlocksMarkersSet != null) {
+					world.get().getMaps().forEach(e -> e.getMarkerSets().remove("townyprovinces.markerset.homeblocks"));
+				}
+				for (BlueMapMap map : world.get().getMaps()) {
+					map.getMarkerSets().put("townyprovinces.markerset.homeblocks", homeBlocksMarkersSet);
+				}
+			}
+			drawProvinceHomeBlocks();
+			drawProvinceBorders();
+			}
+		);
+	}
+
+	
+
+	@Override
+	protected void drawProvinceHomeBlocks() {
+		String border_icon_id = "provinces_costs_icon";
+		boolean biomeCostAdjustmentsEnabled = TownyProvincesSettings.isBiomeCostAdjustmentsEnabled();
+		Set<Province> provinceSet = TownyProvincesDataHolder.getInstance().getProvincesSet();
+
+
+		
+		for(Province province : provinceSet){
+			try{
+				String iconUrl = "assets/province.png";
+				TPCoord homeBlock = province.getHomeBlock();
+				String homeBlockMarkerId = "province_homeblock_" + homeBlock.getX() + "-" + homeBlock.getZ();
+				Marker homeBlockMarker = homeBlocksMarkersSet.get(homeBlockMarkerId);
+				if(province.getType().canNewTownsBeCreated()){
+
+					if(homeBlockMarker != null) 
+						continue;
+					int realHomeBlockX = homeBlock.getX() * TownyProvincesSettings.getChunkSideLength();
+					int realHomeBlockZ = homeBlock.getZ() * TownyProvincesSettings.getChunkSideLength();
+
+						String markerLabel;
+						if (TownyEconomyHandler.isActive()) {
+							int newTownCost = (int) (biomeCostAdjustmentsEnabled ? province.getBiomeAdjustedNewTownCost() : province.getNewTownCost());
+							String newTownCostString = TownyEconomyHandler.getFormattedBalance(newTownCost);
+							int upkeepTownCost = (int) (biomeCostAdjustmentsEnabled ? province.getBiomeAdjustedUpkeepTownCost() : province.getUpkeepTownCost());
+							String upkeepTownCostString = TownyEconomyHandler.getFormattedBalance(upkeepTownCost);
+							markerLabel = Translatable.of("dynmap_province_homeblock_label", newTownCostString, upkeepTownCostString).translate(Locale.ROOT);
+						} else {
+							markerLabel = "";
+						}
+
+						homeBlockMarker = POIMarker.builder()
+							.label(markerLabel)
+							.detail(markerLabel)
+							.position(realHomeBlockX, 65.0, realHomeBlockZ)
+							.icon(iconUrl, 8, 8)
+							.build();
+
+						homeBlocksMarkersSet.put(homeBlockMarkerId, homeBlockMarker);
+
+					}else{
+						
+						if(homeBlockMarker == null)
+							continue;
+						homeBlocksMarkersSet.remove(homeBlockMarkerId);
+						return;
+					}
+			}catch (Exception ex){
+				TownyProvinces.severe("Problem adding homeblock marker");
+				ex.printStackTrace();
+			}
+		}
+	}
+
+	@Override
+	protected void setProvinceMapStyles() {
+		Marker marker;
+		for(Province province : TownyProvincesDataHolder.getInstance().getProvincesSet()){
+			marker = borderMarkerSet.get(province.getId());
```

Comment:
🚧 
To avoid concurrency issue, use `new HashSet(TownyProvincesDataHolder.getInstance().getProvincesSet()))"`